### DISCUSSION TITLE
typing issue with extended client

### DISCUSSION THREAD

**Discussion by yovanoc** [prisma/prisma#20914] [Q&A]
### Question

The need is to use prisma typing for an external function but with extended client.
But I don't manage to get the typing working but maybe there is a better way because you will see in my code snippet a "hacky" thing to get the `ExtArg` but anyway this is not working overall. 

I take reference from the findFirst method of this model in the generated typings.

Happy to see if there is better (working) way to do this.

### How to reproduce (optional)

Giving this schema:

```prisma
model FirstTable {
  id        String     @id @default(cuid())
  secondTable SecondTable?
}

model SecondTable {
  id                 String       @id @default(cuid())
  firstTable    FirstTable @relation(fields: [firstTableId], references: [id])
  firstTableId String       @unique
}
```

and my try to do this kind of function:

```ts
import type { Prisma } from '@prisma/client';
import { type GetResult } from '@prisma/client/runtime/library';

import { prisma } from '~/server/db';

type ExtArgs = (typeof prisma)['$extends']['extArgs'];

type FirstTableFindArgs = Omit<
  Prisma.FirstTableFindFirstArgs<ExtArgs>,
  'where'
>;

export function myFunction<T extends Prisma.FirstTableFindFirstArgs<ExtArgs>>(
  {
    id,
  }: {
    id: string;
  },
  args: Prisma.SelectSubset<T, FirstTableFindArgs>
): Prisma.Prisma__FirstTableClient<
  GetResult<Prisma.$FirstTablePayload<ExtArgs>, T, 'findFirst'> | null,
  null,
  ExtArgs
> {
  return prisma.firstTable.findFirst({
    ...args,
    where: {
      id: id,
    },
  });
}
```

I have the following error: 

```ts
Type 'DynamicModelExtensionFluentApi<TypeMap<Args_2 & { result: {}; model: { $allModels: { createSelect: () => <T, const A extends Args_3<T, "findUniqueOrThrow">["select"]>(this: T, args: A) => A; }; user: { createSelect: () => <T, const A extends Args_3<...>["select"]>(this: T, args: A) => A; }; ... 46 more ...; appLimit...' is not assignable to type 'Prisma__FirstTableClient<GetFindResult<$FirstTablePayload<{ result: {}; model: { $allModels: { createSelect: () => <T, const A extends Args_3<T, "findUniqueOrThrow">["select"]>(this: T, args: A) => A; }; user: { createSelect: () => <T, const A extends Args_3<T, "findUniqueOrThrow">["select"]>(this: T, args: A) =...'.
  Types of property 'secondTable' are incompatible.

....
```

my extended prisma client was originally created like this:

```ts
import { type Prisma, PrismaClient } from '@prisma/client';

const transform = (prisma: PrismaClient) =>
  prisma.$extends({
    model: {
      $allModels: {
        createSelect<
          T,
          const A extends Prisma.Args<T, 'findUniqueOrThrow'>['select'],
        >(this: T, args: A): A {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-return
          return args;
        },
      },
    },
  });

export type SecrecyDbClient = ReturnType<typeof transform>;

const globalForPrisma = globalThis as { prisma?: SecrecyDbClient | undefined };

export const prisma = globalForPrisma.prisma ?? transform(new PrismaClient());

if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma;
```

### Information about Prisma Schema, Client Queries and Environment (optional)

macOs
postgres
node 20.5.1

```
prisma                  : 5.2.0
@prisma/client          : 5.2.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 2804dc98259d2ea960602aca6b8e7fdc03c1758f (at node_modules/.pnpm/@prisma+engines@5.2.0/node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Schema Engine           : schema-engine-cli 2804dc98259d2ea960602aca6b8e7fdc03c1758f (at node_modules/.pnpm/@prisma+engines@5.2.0/node_modules/@prisma/engines/schema-engine-darwin)
Schema Wasm             : @prisma/prisma-schema-wasm 5.2.0-25.2804dc98259d2ea960602aca6b8e7fdc03c1758f
Default Engines Hash    : 2804dc98259d2ea960602aca6b8e7fdc03c1758f
Studio                  : 0.494.0
Preview Features        : fullTextSearch, metrics
```

**Comment by millsp** [comment]
You can re-use your extended Prisma Client types this way (and it also works for your regular client). I wanted to recommend you to do something like this, where `prisma` is your extended client...

```ts
function myFunction<const T extends Prisma.Args<typeof prisma.user, 'findFirst'>>(
    { id }: { id: string },
    args: T,
) {
    return prisma.user.findFirst({
        ...args,
        where: { id },
    });
}
```

...but that won't work, although it should, I have uncovered a type-level bug with our `Prisma.Exact` types which are also used internally.

**Reply by millsp** [reply]
Sorry about that, in the mean time I have created an issue https://github.com/prisma/prisma/issues/20949

**Reply by millsp** [reply]
In the mean time, I have this ugly solution for you

```ts
function myFunction<T extends Prisma.Args<typeof prisma.user, 'findFirst'>>(
    { id }: { id: string },
    args: Prisma.Exact<T, Prisma.Args<typeof prisma.user, 'findFirst'>>
) {
    args.where = { id }
    return prisma.user.findFirst(args);
}
```

**Reply by yovanoc** [reply]
Thanks for this but the main issue was to typed the return type

**Reply by gurpreetatwal** [reply]
@yovanoc were you able to get a solution for the return type? I'm stuck in a similar situation. I've tried a bunch of combinations with `Prisma.Result` and am not able to get it to work

Based on my understanding of the docs the following should work, but I can't get it to


```typescript
export async function query<
  const T extends PrismaT.Args<typeof prisma.study, 'findUniqueOrThrow'>,
>(
  ctx: { organization: { id: string } },
  studyId: string,
  args: Omit<T, 'where'>,
): Promise<PrismaT.Result<typeof prisma.study, T, 'findUniqueOrThrow'>> {
  return await prisma.study.findUniqueOrThrow({
    ...args,
    where: {
      id: studyId,
      organizationId: ctx.organization.id,
    },
  });
}
```

I get the following error on the `return` statement: 

```
Type '{ id: string; organizationId: ... 12 more ...; updatedAt: Date; }' is not assignable to type 'GetFindResult<$StudyPayload<DefaultArgs>, T>
```

**Reply by gurpreetatwal** [reply]
I was able to somewhat get it working!! The key was two-fold:
1. Wrap the `args` parameter in `SelectSubset` 
2. Give up on omitting the where clause

```typescript
export async function query<T extends PrismaT.Args<typeof prisma.study, 'findUniqueOrThrow'>>(
  ctx: { profile: { id: string }; organization: { id: string } },
  studyId: string,
  args: PrismaT.SelectSubset<T, PrismaT.Args<typeof prisma.study, 'findUniqueOrThrow'>>,
) {
  args.where = {
    id: studyId,
    organizationId: ctx.organization.id,
  };

  return await prisma.study.findUniqueOrThrow(args);
}
```
(also it's possible to use `PrismaT.StudyFindUniqueOrThrowArgs` instead of `PrismaT.Args<typeof prisma.study, 'findUniqueOrThrow'>` to make the code a bit more concise)

This results in the return type correctly having the selected fields or included relations at the cost of having to pass a dummy where clause in the input. I hope fixing #20949 will allow for omitting the where clause, but don't know enough to be sure if the two issues are related or not.

**Reply by yovanoc** [reply]
Yes, my use case was to have the where clause constructed by my function's inputs so I did something similar

**Reply by gurpreetatwal** [reply]
Were you able to resolve the issue of having to pass the where clause specifically passed as an argument? I currently have to pass `{where: {id: 'dummy'}}` in order to satisfy typescript

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
